### ISSUE TITLE
Get the actual value

### ISSUE THREAD

**Issue by sagikazarmark** [thenable/promise#3]
Hey there,

I am the project lead for PHP-HTTP/HTTPlug. We are working on an HTTP Client abstraction which supports async requests.

We have the following Promise interface:

https://github.com/php-http/promise/blob/master/src/Promise.php

The idea is to replace this with a (hopefully) upcoming PSR.

Honestly this thenable interface looks very good, although not sure if it fits for our needs.

Currently we heavily rely on a `wait` method as our plugin system uses Promises and emulates async requests when needed. Also emulates sync requests from async ones to avoid code duplication in some cases.

So maybe you could give me a hint if we could use this interface, extend it, whatever.

From a standardizing point of view: not sure if done and except should be part of the main interface. Isn't just then enough? In guzzle's similar implementation these are just convenience methods and internally call `then` anyway. So my question is: are there any reasons why these methods must be in the main interface instead of an ExtendedPromise or a MuchBetterNamedPromise interface?

**Comment by trowski** [comment]
Icicle has a [`wait`](https://github.com/icicleio/icicle/blob/master/src/Awaitable/Awaitable.php#L130) method on [Awaitable](https://github.com/icicleio/icicle/blob/master/src/Awaitable/Awaitable.php) (or Promise) that does exactly what you're describing, stoping control-flow at the method call until the promise is resolved. Maybe this could be part of the interface as a method for mixing async code with sync code (thought it will have to come with a big warning in the method description).

Technically you can implement `wait` as a function using `done` (or `then`), but it's not terribly elegant (this code is based on Icicle, using other libraries would vary):

```php
function wait(Awaitable $awaitable) {
    $resolved = false;
    $awaitable->done( // then() also would work.
        function ($value) use (&$resolved, &$result, &$exception) {
            $resolved = true;
            $value = $value;
        },
        function (\Exception $reason) use (&$resolved, &$result, &$exception) {
            $resolved = true;
            $exception = $reason;
        }
    );

    while (!$resolved) {
        Loop\tick(); // Manually tick the event loop.
    }

    if ($exception) {
        throw $exception;
    }

    return $result;
}
```

---

`done` is part of the interface to provide a more performant way to invoke functions that consume the result or handle an error rather than transforming the value. Returning another promise is wasteful if you don't plan on returning a value from the callback functions. This is particularly relevant with [coroutines](https://github.com/icicleio/icicle/blob/master/src/Coroutine/Coroutine.php) that use promises to define interruption points and simply use the resolution of the promise to continue execution of the coroutine.

`except` was included in the interface for chaining with `then`, but is not strictly needed. The name was chosen to avoid conflicting with existing implementations that chose a different name. For a standard interface it may be better to omit the method altogether.

**Comment by sagikazarmark** [comment]
Thanks for your response @trowski. Your explanation helped me a lot.

Based on the other conversation about the naming, I thought this is going to be some sort of proposal to the FIG. Am I wrong about it?

Nevertheless, I would be happy to see any kind of Promise contract, even if it is not inside the FIG.

About the wait method: I think it would make sense to include it in the interface, although I am not completely sure if it should accept a Promise as the argument. Is it part of the A+ spec?

To tell the truth, I like how guzzle does this:

https://github.com/guzzle/promises/blob/master/src/PromiseInterface.php#L92

At least it makes sense to me to make unwrapping the value optional while you can still wait for it and call the underlying callbacks. WDYT?

Or should it be in an unwrap method? Honestly, I haven't read closely the spec.

**Comment by trowski** [comment]
Yes, the purpose of this repo is to act as a basis for a proposal to FIG, however it doesn't mean that *this* repo has to conform exactly to their naming conventions since the FIG would never officially use this repo.

If a `wait` method was added it would be included in the interface. I was just demonstrating how a function could be implemented to do something similar using just `done` or `then`. The [A+ spec](https://promisesaplus.com) does not include a `wait` method, only `then`. However, I think it would be beneficial to add a `done` and perhaps a `wait` method.

In my opinion the `wait` method should be similar to the `wait` method on Icicle's Awaitables: https://github.com/icicleio/icicle/blob/master/src/Awaitable/Awaitable.php#L130

Why should unwrapping the value with `wait` be optional? Why would you stop control flow to wait for a promise to resolve if you didn't need the resolution value at that point in the code? Is there a use case I'm not seeing?

**Comment by sagikazarmark** [comment]
Maybe I miss something as well. I thought thought the done callbacks can be used for processing the returned value. In some cases you might not need the value, only in those callbacks.

Actually I would be happy with a solution when some methods are separated into separate interfaces. For example a Promise with a wait method could actually be an Awaitable interface extending the Promise interface. Same could be done with done. This way the promise itself can be kept clean. The downside is that you have to do type checks when using promises.

**Comment by sagikazarmark** [comment]
Anyway, as I linked above, we are currently maintaining our own interface. If you plan moving on with yours as a standard, I would happily use it if it gets sort of stable, before HTTPlug does.

**Comment by trowski** [comment]
I plan on updating this over the weekend. Hopefully you can make use of this in HTTPlug.

I think `Thenable` would be the most appropriate name for the interface.

The interface will need to have `then()` and `done()` at a minimum. I'm still conflicted about `except()` and `wait()`.

`except()` isn't strictly necessary, it's just a convenience method with an obvious implementation. I like the idea of making it like Icicle's `capture()` method or React's `otherwise()` method that examines the type declaration on the passed callable and invokes the function only if the rejection reason matches the type, but this requires reflection. I'm not sure I want to include something in an interface that requires reflection to implement.

`wait()` is useful when you want to move from asynchronous code to synchronous code, but I'm concerned about making it a part of the interface, since libraries like React that do not have a dependency on the event loop might have a difficult time implementing the method. As I showed above, there are ways to implement `wait()` using only `done()` that can vary depending on the event loop used.

Thoughts?

**Comment by sagikazarmark** [comment]
That's why I proposed having a separate interface with having the wait method. Thenable and Promise?

I think if `wait` is necessary, we need that as part of the contract.

**Comment by trowski** [comment]
I don't see the value in a separate interface. What happens if you get a `Thenable` but you want to synchronously wait for it to resolve? An application would have to handle that situation as well. So either the interface should include a `wait()` method or not (and I think not because it would tie the implementing object to an event loop or some other resolving method).

**Comment by sagikazarmark** [comment]
Usually promises are returned from a library, aren't they? Because in this case it depends on the library what is returned, which is predictable. Httplug would return a Promise. React would return a thenable.

Having a wait method does not mean it has to be functional. But it would still depend on the library you use it in.

With other words: the behavior of the promise depends on the library you return it from. That's why I think having two interfaces would be no big deal: either separate or one interface, the application still has to handle it according to the library.

**Comment by kelunik** [comment]
I don't think `wait` should be part of the interface. Waiting isn't the responsibility of a promise.
